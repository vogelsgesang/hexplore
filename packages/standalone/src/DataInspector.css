.hv-data-inspector {
    padding: .4em;
}

.hv-data-inspector-caption {
    font-size: .8em;
    margin-top: .6em;
    color: var(--secondary-placeholder-fg-color);
    text-align: right;
}

.hv-data-inspector-value {
    border: 1px solid var(--secondary-border-color);
    padding: .1em;
    text-align: right;
}