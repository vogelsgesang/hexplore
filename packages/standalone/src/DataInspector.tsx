import React, {useMemo} from "react";
import {createRenderer, humanReadableRendererName, RendererConfig} from "hexplore-hexview/dist/ByteRenderer";
import {createAddressRendererConfig, createIntegerRendererConfig} from "hexplore-hexview/dist/ByteRenderer";

import "./DataInspector.css";

export const defaultInspectorRepresentations: RendererConfig[] = [
    createAddressRendererConfig(),
    createIntegerRendererConfig({displayBase: 10, width: 1, signed: true, fixedWidth: false}),
    createIntegerRendererConfig({displayBase: 10, width: 2, signed: true, fixedWidth: false}),
    createIntegerRendererConfig({displayBase: 10, width: 4, signed: true, fixedWidth: false}),
    createIntegerRendererConfig({displayBase: 10, width: 8, signed: true, fixedWidth: false}),
];

interface DataInspectorProps {
    data: ArrayBuffer;
    position: number;
    representations: RendererConfig[];
}

export function DataInspector({data, position, representations}: DataInspectorProps) {
    const view = useMemo(() => new DataView(data), [data]);
    return (
        <div className="hv-data-inspector">
            {representations.map((r, i) => (
                <DataInspectorCell key={i} data={view} position={position} representation={r} />
            ))}
        </div>
    );
}

interface DataInspectorCellProps {
    data: DataView;
    position: number;
    representation: RendererConfig;
}

function DataInspectorCell({data, position, representation}: DataInspectorCellProps) {
    const renderer = useMemo(() => createRenderer(representation), [representation]);
    return (
        <React.Fragment>
            <div className="hv-data-inspector-caption">{humanReadableRendererName(representation)}</div>
            <div className="hv-data-inspector-value">{renderer(data, position)}</div>
        </React.Fragment>
    );
}
