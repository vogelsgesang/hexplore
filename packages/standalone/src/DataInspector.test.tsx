import React from "react";
import snapshotRenderer from "react-test-renderer";
import {DataInspector, defaultInspectorRepresentations} from "./DataInspector";

function constData(bytes: number[]) {
    const buf = new ArrayBuffer(bytes.length);
    const bufView = new Uint8Array(buf);
    for (let i = 0, len = bytes.length; i < len; i++) {
        bufView[i] = bytes[i];
    }
    return buf;
}

describe("displays the data correctly", () => {
    const testData = constData([1, 2, 3, 4, 5, 6, 7, 8]);
    function testRendering(pos: number) {
        const component = snapshotRenderer.create(
            <DataInspector data={testData} position={pos} representations={defaultInspectorRepresentations} />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    }

    test("for address 0", () => {
        testRendering(0);
    });
    test("for address 1", () => {
        testRendering(1);
    });
    test("for last address", () => {
        testRendering(testData.byteLength - 1);
    });
});
