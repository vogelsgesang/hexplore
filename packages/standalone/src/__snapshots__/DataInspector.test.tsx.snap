// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the data correctly for address 0 1`] = `
<div
  className="hv-data-inspector"
>
  <div
    className="hv-data-inspector-caption"
  >
    Address
  </div>
  <div
    className="hv-data-inspector-value"
  >
    0x0
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 1-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     1
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 2-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     513
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 4-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     67305985
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 8-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     578437695752307201
  </div>
</div>
`;

exports[`displays the data correctly for address 1 1`] = `
<div
  className="hv-data-inspector"
>
  <div
    className="hv-data-inspector-caption"
  >
    Address
  </div>
  <div
    className="hv-data-inspector-value"
  >
    0x1
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 1-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     2
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 2-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     770
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 4-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     84148994
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 8-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
    .
  </div>
</div>
`;

exports[`displays the data correctly for last address 1`] = `
<div
  className="hv-data-inspector"
>
  <div
    className="hv-data-inspector-caption"
  >
    Address
  </div>
  <div
    className="hv-data-inspector-value"
  >
    0x7
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 1-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
     8
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 2-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
    .
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 4-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
    .
  </div>
  <div
    className="hv-data-inspector-caption"
  >
    Signed 8-byte Decimal
  </div>
  <div
    className="hv-data-inspector-value"
  >
    .
  </div>
</div>
`;
